content/blog/tidymodels-2022-q3/index.Rmd

```diff
@@ -0,0 +1,248 @@
+```
+
+then pass this to any of the existing control functions: 
+
+```r
+grid_ctrl <- control_grid(backend_options = h2o_thread_spec)
+```
+
+Now h2o can parallel process 10 models at once. 
+
+Here is an example using a simulated data set with a numeric outcome: 
+
+```{r}
+#| results: hide
+
+library(tidymodels)
+library(agua)
+library(finetune)
+
+# Simulate the data
+n_train <- 200
+set.seed(6147)
+sim_dat <- sim_regression(n_train)
+
+# Resample using 10-fold cross-validation
+set.seed(91)
+sim_rs <- vfold_cv(sim_dat)
+```
+
+We'll use grid search to tune a boosted tree:
+
+```{r}
+boost_spec <-
+  boost_tree(
+    trees = tune(),
+    min_n = tune(),
+    tree_depth = tune(),
+    learn_rate = tune(),
+    loss_reduction = tune()
+  ) %>%
+  set_engine("h2o") %>%
+  set_mode("regression")
+```
+
+Now let's parallel process our computations
+
+```{r}
+#| results: hide
+#| label: h2o-init
+
+# Start the h2o server
+h2o::h2o.init()
+
+# Multi-thread the model fits
+h2o_thread_spec <- agua_backend_options(parallelism = 10)
+grid_ctrl <- control_grid(backend_options = h2o_thread_spec)
+```
+
+We'll evaluate a very small grid at first: 
+
+```{r}
+#| results: hide
+#| message: FALSE
+#| warning: FALSE
+#| label: grid
+#| cache: TRUE
+set.seed(7616)
+grid_res <-
+  boost_spec %>%
+  tune_grid(outcome ~ ., resamples = sim_rs, grid = 10, control = grid_ctrl)
+```
+
+```{r}
+#| label: grid-plot
+#| out.width: "90%"
+#| fig.align: "center"
+#| fig.width: 8
+#| fig.height: 5
+#| dev: svg
+show_best(grid_res, metric = "rmse") %>% select(-.config, -.metric, -.estimator)
+
+autoplot(grid_res, metric = "rmse")
+```
+
+It was a small grid and most of the configurations were not especially good. We can further optimize the results by applying simulated annealing search to the best grid results. 
+
+```{r}
+#| label: sim-anneal
+#| cache: TRUE
+sa_ctrl <- control_sim_anneal(backend_options = h2o_thread_spec)
+
+set.seed(4)
+sa_res <-
+  boost_spec %>%
+  tune_sim_anneal(
+    outcome ~ .,
+    resamples = sim_rs,
+    initial = grid_res,
+    iter = 25,
+    control = sa_ctrl
+  )
+```
+
+Again, h2o is doing all of the computational work for fitting models and tidymodels is proposing new parameter configurations. 
+
+One other nice feature of the new agua release is the h2o engine for the `auto_ml()` model. This builds a stacked ensemble on a set of different models (not unlike our [stacks](https://stacks.tidymodels.org) package but with far less code). 
+
+There is a great worked example [on the agua website](https://agua.tidymodels.org/articles/auto_ml.html) so make sure to check this out! 
+
+## More spline recipe steps
+
+Spline techniques allow linear models to produce nonlinear model curves These are called [basis expasion methods](https://bookdown.org/max/FES/numeric-one-to-many.html#numeric-basis-functions) since they take a single numeric predictor and make additional nonlinear feature columns. 
```

Comment:
```suggestion
Spline techniques allow linear models to produce nonlinear model curves. These are called [basis expasion methods](https://bookdown.org/max/FES/numeric-one-to-many.html#numeric-basis-functions) since they take a single numeric predictor and make additional nonlinear feature columns. 
```